include/nlohmann/json.hpp

```diff
@@ -2146,6 +2185,9 @@ class basic_json
     basic_json(basic_json&& other) noexcept
         : m_type(std::move(other.m_type)),
           m_value(std::move(other.m_value))
+#if JSON_DIAGNOSTICS
+        , m_parent(other.m_parent)
```

Comment:
This gives the user an opportunity to create an object with a dangling pointer. For example:
```c++
json j = std::move(json{0}[0]); // extract nested object from json rvalue
```
The dangling pointer leads to `SIGSEGV` when `diagnostics()` is called (see also https://godbolt.org/z/jKGsGc).
The object being initialized by move constructor doesn't actually reside inside of `*other.m_parent`. The latter still contains the `other` (until destruction) and has no relation to the new object, which should have no parent in its initial state.

Reply:
Man that's ugly code, but agreed that the constructors shouldn't be setting the parent to anything other than `nullptr`, which means it can just use the NSDMI in the definition.

Reply:
What do you propose as fix?

Reply:
1. As @gregmarr said, the NSDMI in the class definition will be enough here. The above line needs just to be deleted. After the object is initialized, the parent must be `nullptr`.
2. Taking into account the previous, the assignment below should be performed after the element is moved into a container (as also proposed by @gregmarr). https://github.com/nlohmann/json/blob/294fa343d54634856dfef4a645de5241bd2dcde1/include/nlohmann/json.hpp#L1646
3. The `operator = ` should not modify the parent pointer.